crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,483 @@
+///
+///   /// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_num_instructions(qc); // 1
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_num_instructions(circuit: *const CircuitData) -> usize {
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    circuit.__len__()
+}
+
+#[repr(C)]
+pub struct CInstruction {
+    name: *mut c_char,
+    num_qubits: u32,
+    qubits: *mut u32,
+    num_clbits: u32,
+    clbits: *mut u32,
+    num_params: u32,
+    params: *mut f64,
+}
+
+/// @ingroup QKCircuit
+/// Return the number of instructions in the circuit
+///
+/// @param m circuit A pointer to the circuit to get the counts for.
+///
+///   /// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_get_instruction(qc, 0); //
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_get_instruction(
+    circuit: *const CircuitData,
+    index: usize,
+) -> CInstruction {
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    if index >= circuit.__len__() {
+        panic!("Invalid index")
+    }
+    let packed_inst = &circuit.data()[index];
+    let qargs = circuit.get_qargs(packed_inst.qubits);
+    let mut qargs_vec: Vec<u32> = qargs.iter().map(|x| x.0).collect();
+    let cargs = circuit.get_cargs(packed_inst.clbits);
+    let mut cargs_vec: Vec<u32> = cargs.iter().map(|x| x.0).collect();
+    let params = packed_inst.params_view();
+    let mut params_vec: Vec<f64> = params
+        .iter()
+        .map(|x| match x {
+            Param::Float(val) => *val,
+            _ => unreachable!("Invalid parameter on instruction"),
+        })
+        .collect();
+    let out_qargs = qargs_vec.as_mut_ptr();
+    std::mem::forget(qargs_vec);
+    let out_cargs = cargs_vec.as_mut_ptr();
+    std::mem::forget(cargs_vec);
+    let out_params = params_vec.as_mut_ptr();
+    std::mem::forget(params_vec);
+
+    CInstruction {
+        name: CString::new(packed_inst.op.name()).unwrap().into_raw(),
+        num_qubits: qargs.len() as u32,
+        qubits: out_qargs,
+        num_clbits: cargs.len() as u32,
+        clbits: out_cargs,
+        num_params: params.len() as u32,
+        params: out_params,
+    }
+}
+
+/// @ingroup QkCircuit
+/// Free a circuit instruction object
+///
+/// @param inst The instruction to free
+///
+/// # Safety
+/// Behavior is undefined if ``inst`` is not an object returned by ``qk_circuit_get_instruction``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_circuit_instruction(inst: CInstruction) {
```

Comment:
This should be `qk_circuit_instruction_free` for consistency -- and should it take a pointer to a `CInstruction` otherwise the struct would get passed by value?

Reply:
But don't we also need to free the struct if it was allocated by `qk_circuit_get_instruction`? 🤔

Reply:
It's a directly returned stack allocated struct so we don't have to worry about freeing it, the memory will be reclaimed automatically like any local variable (unless the caller puts it on the heap but then it's their responsibility).

Reply:
Ah right because we don't Box::free anything we don't have to consume it again -- like the 1:1 mapping of `malloc` and `free` in C